File: codes/semana11/Singleton.cpp
// -------------------------------------------
// -------------------------------------------

#include<string>
#include<iostream>

using namespace std;

// -------------------------------------------
// -------------------------------------------

class Singleton {
    public:
        static Singleton* Instance();
        string getName();
        
    protected:
        Singleton(string);

    private:
        static Singleton* _instance;
        string name;
};

// -------------------------------------------
// -------------------------------------------

Singleton* Singleton::_instance = 0;

// -------------------------------------------
// -------------------------------------------

Singleton::Singleton(string n) {
    name = n;
}

// -------------------------------------------
// -------------------------------------------

Singleton* Singleton::Instance() {
    if(_instance == 0) {
        _instance = new Singleton("unico");
    }
    return (_instance);
}

// -------------------------------------------
// -------------------------------------------

string Singleton::getName() {
    return(name);
}

// -------------------------------------------
// -------------------------------------------

int main(int argc, char* argv[]) {

    Singleton* a = Singleton::Instance();
    Singleton* b = Singleton::Instance();

    cout << "nome de A: " << a->getName() << endl;
    cout << "nome de B: " << b->getName() << endl;
}

// -------------------------------------------
// -------------------------------------------

File: codes/semana11/Singleton.py
class Singleton:

    _instance = None

    def __init__(self, name: str):
        self.name = name

    @classmethod
    def Instance(cls):
        if cls._instance is None:
            cls._instance = Singleton("unico")
        return cls._instance

    def getName(self) -> str:
        return self.name


if __name__ == "__main__":
    a = Singleton.Instance()
    b = Singleton.Instance()
    c = Singleton.Instance()
    d = Singleton.Instance()

    print(f"nome de A: {a.getName()}")
    print(f"nome de B: {b.getName()}")
    print(f"nome de C: {c.getName()}")
    print(f"nome de D: {d.getName()}")
